[12.0][IMP] - Add a company config option to decide whether to create or to extend contract line at renew action

sbejaoui opened the pull request:
Add a company config option to decide whether to create or to extend contract
line at renew action

sbidoul commented:
@sbejaoui We can probably avoid a breaking change by adding an option on contract l ine that controls how the renewal is done: by adding a new line (current behavior) or by changing the end date.

sbejaoui commented:
> @sbejaoui We can probably avoid a breaking change by adding an option on contract l ine that controls how the renewal is done: by adding a new line (current behavior) or by changing the end date.

ok, good idea

sbejaoui committed:
[12.0][IMP] - add new option: create_new_line_at_contract_line_renew

Add a company config option to decide whether to create or to extend contract
line at renew action

sbejaoui committed:
[12.0][IMP] - extend contract line at renewal

sbejaoui commented:
cc @gva-acsone

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -1164,17 +1164,33 @@ def _get_renewal_dates(self):
+            if company.create_new_line_at_contract_line_renew:
+                new_line = rec._renew_create_line(date_start, date_end)
+            else:
+                new_line = rec._renew_extend_line(date_end)
```
It bothers me that one method takes `date_start` and the other not. It should be symmetric or bugs will creep in later. Since it does not make sense for `_renew_extend_line` to take `date_start` then it must be removed from `_renew_create_line` (in which case it is always date_end + 1 day), and `_get_renewal_dates` must be changed to `_get_renewal_new_date_end`.

sbejaoui replied:
make sense

sbejaoui committed:
[IMP] - improve code: unify methods argument _renew_create_line and _renew_extend_line

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -1156,25 +1156,43 @@ def action_stop_plan_successor(self):
-            new_line._onchange_date_start()
+            company = rec.contract_id.company_id
+            date_end = rec._get_renewal_new_date_end()
+            date_start = rec.date_end + relativedelta(days=1)
```
```suggestion
```

sbejaoui replied:
it's used in the contract posted message

https://github.com/OCA/contract/blob/c15a876603df32c82bd051df95015c04c2939f1e/contract/models/contract_line.py#L1197-L1209

sbidoul replied:
Ah, yes, sorry.

sbidoul reviewed (approved):
LGTM

gva-acsone reviewed (approved):
LGTM Functional test

sbejaoui commented:
/ocabot merge patch
